### PR TITLE
feature: support regex_uri option in proxy-rewrite plugin

### LINKED ISSUE #949: request help: How to replace uri?
route configures:
urls: /abc
upstream: abc   

upstream configures:
abc: 127.0.0.1:17777

want:
curl 127.0.0.1:9080/abc/hello -> 127.0.0.1:17777/hello?

### PR DESCRIPTION
**Pull request by agile6v**
### Summary

Support more flexible URI rewriting in proxy-rewrite plugin. 

Using regex_uri option to support regex matching. Its type is array, the first element is the matching pattern, and the second element is the final target. If the match is successful, the second element will be used as the URI that is forwarded to the upstream. If uri and regex_uri option are configured at the same time, then uri will be used first.

For example:

```shell
{
    "methods": ["GET"],
    "uri": "/abc/def/*",
    "plugins": {
        "proxy-rewrite": {
            "regex_uri": ["^/abc/def/(.*)", "/$1"]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:80": 1
        }
    }
}
```

### Issues resolved

Fix #949

### CODE REVIEW

**Comment by juzhiyuan** [comment]
Could anyone help to review this PR?

**Comment by moonming** [comment]
> Could anyone help to review this PR?

On my way :)

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 1]
```diff
@@ -31,6 +31,16 @@ local schema = {
             minLength   = 1,
             maxLength   = 4096
         },
+        regex_uri = {
+            description = "new uri for upstream, lower priority than uri property",
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: I think it's a substitution, not set a new uri, right?

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 2]
```diff
@@ -110,7 +120,18 @@ function _M.rewrite(conf, ctx)
         end
     end
 
-    local upstream_uri = conf.uri or ctx.var.uri
+    local upstream_uri = ctx.var.uri
+    if conf.uri ~= nil then
+        upstream_uri = conf.uri
+    elseif conf.regex_uri ~= nil then
+        local uri, _, err = ngx.re.sub(ctx.var.uri, conf.regex_uri[1], conf.regex_uri[2], "o")
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: you should enable PCRE JIT compilation as 'jo' not only 'o'

Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: better use `local re_sub = ngx.re.sub`

File: t/plugin/proxy-rewrite.t [hunk 3]
```diff
@@ -743,3 +743,141 @@ x-api-engine: APISIX
...
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "methods": ["GET"],
+                        "plugins": {
+                            "proxy-rewrite": {
+                                "regex_uri": ["^/test/(.*)", "/$1"]
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: only support `$1`, right?

File: t/plugin/proxy-rewrite.t [hunk 4]
```diff
@@ -743,3 +743,141 @@ x-api-engine: APISIX
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 27: rewrite uri using regex_uri
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: need to add one 404 test case.

File: t/plugin/proxy-rewrite.t [hunk 5]
```diff
@@ -743,3 +743,141 @@ x-api-engine: APISIX
...
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 29: rewrite uri using uri & regex_uri property
+--- request
+GET /test/hello HTTP/1.1
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: `/test/hello` will both match `uri` and `regex_uri`, so it's not a good test case.

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 6]
```diff
@@ -110,7 +120,18 @@ function _M.rewrite(conf, ctx)
...
+        if uri then
+            upstream_uri = uri
+        else
+            core.log.error("failed to substitute the uri[", ctx.var.uri, "]:", err)
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: need a test case for this line.

**Comment by agile6v** [comment]
Thanks for your suggestion. I'll fix them. @moonming

**Comment by moonming** [comment]
@agile6v please add the document, thx

File: t/plugin/proxy-rewrite.t [hunk 7]
```diff
@@ -743,3 +743,198 @@ x-api-engine: APISIX
...
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "methods": ["GET"],
+                        "plugins": {
+                            "proxy-rewrite": {
+                                "regex_uri": ["^/test/(.*)", "/$`1"]
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: we should  check whether the `regex_uri` is valid in `check_schema` function.

Comment (moonming) [reply t/plugin/proxy-rewrite.t]: There is one issue left, else are good to me.

File: t/plugin/proxy-rewrite.t [hunk 8]
```diff
@@ -743,3 +743,198 @@ x-api-engine: APISIX
...
+
+=== TEST 26: set route(only using regex_uri)
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "methods": ["GET"],
```
Comment (membphis) [inline t/plugin/proxy-rewrite.t]: This is not a required option, we can delete it.

File: t/plugin/proxy-rewrite.t [hunk 9]
```diff
@@ -743,3 +743,198 @@ x-api-engine: APISIX
...
+
+=== TEST 29: set route(Using both uri and regex_uri)
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "methods": ["GET"],
```
Comment (membphis) [inline t/plugin/proxy-rewrite.t]: ditto

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 10]
```diff
@@ -110,7 +122,21 @@ function _M.rewrite(conf, ctx)
...
+    local upstream_uri = ctx.var.uri
+    if conf.uri ~= nil then
+        upstream_uri = conf.uri
+    elseif conf.regex_uri ~= nil then
+        local uri, _, err = re_sub(ctx.var.uri, conf.regex_uri[1],
+                                   conf.regex_uri[2], "jo")
+        if uri then
+            upstream_uri = uri
+        else
+            core.log.error("failed to substitute the uri ", ctx.var.uri,
```
Comment (membphis) [inline lua/apisix/plugins/proxy-rewrite.lua]: If something goes wrong here, we should respond directly with code `500`.

File: t/plugin/proxy-rewrite.t (hunk 7 above)
Comment (agile6v) [reply t/plugin/proxy-rewrite.t]: OK.  What I can think of is to execute `ngx.re.sub` in check_schema function at the moment. Do you have any other suggestions?

File: doc/plugins/proxy-rewrite.md [hunk 11]
```diff
@@ -35,6 +35,7 @@ upstream proxy info rewrite plugin.
 |-------         |-----|------|
 |scheme          |optional| Upstream new `schema` forwarding protocol,options can be `http` or `https`,default `http`.|
 |uri             |optional| Upstream new `uri` forwarding address.|
+|regex_uri       |optional| Upstream new `uri` forwarding address. Use regular expression to match uri from client, when the match is successful, the uri template will be forwarded upstream. If the match is not successful, the uri from the client will be forwarded to the upstream. When `uri` and` regex_uri` are both exist, `uri` is used first. For example: [" ^/iresty/(.*)/(.*)/(.*)", "/$1-$2-$3"], the first element represents the matching regular expression and the second element represents the uri template that is forwarded to the upstream.|
```
Comment (membphis) [inline doc/plugins/proxy-rewrite.md]: > If the match is not successful, the uri from the client will be forwarded to the upstream.

If the match fails, is it expected to continue using the original request uri? What do you think of this?

Comment (agile6v) [reply doc/plugins/proxy-rewrite.md]: Similar to this situation in Nginx.
```nginx
server {
	listen 80;
	location / {
	    rewrite ^/hello/(.*) /$1 break;
	    proxy_pass http://127.0.0.1:8080;
	}
}

server {
	listen 8080;
	location / {
		return 200, $uri;
	}
}

$curl http://127.0.0.1/hello/world
Output: /world

$curl http://127.0.0.1/other/things
Output: /other/things
```

**Comment by agile6v** [comment]
Is it better that `uri` option support both types, `string` and `array`?

@moonming @membphis

File: doc/plugins/proxy-rewrite.md (hunk 11 above)
Comment (membphis) [reply doc/plugins/proxy-rewrite.md]: thank you for the explanation.

**Comment by membphis** [comment]
> Is it better that `uri` option support both types, `string` and `array`?
> 
> @moonming @membphis

When do we need `uri` to be an array type?

**Comment by agile6v** [comment]
> > Is it better that `uri` option support both types, `string` and `array`?
> > @moonming @membphis
> 
> When do we need `uri` to be an array type?

I mean is it better to combine `uri` and `regex_uri` into one.

**Comment by agile6v** [comment]
Please review this PR. Thanks. 
@shuaijinchao

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 12]
```diff
@@ -82,6 +82,16 @@ function _M.check_schema(conf)
         return false, err
     end
 
+    if conf.regex_uri and #conf.regex_uri > 0 then
+        local _, _, err = re_sub("/fake_uri", conf.regex_uri[1],
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: what is `/fake_uri`? why we need it?

Comment (agile6v) [reply lua/apisix/plugins/proxy-rewrite.lua]: The `/fake_uri` is used to verify the validity of regex expression.

Comment (moonming) [reply lua/apisix/plugins/proxy-rewrite.lua]: need a test case to cover invalid regex.

Comment (membphis) [reply lua/apisix/plugins/proxy-rewrite.lua]: https://github.com/apache/incubator-apisix/pull/979/files#diff-0dd1fb219e34e914ab6423384290b98cR895

Is this test case?

Comment (moonming) [reply lua/apisix/plugins/proxy-rewrite.lua]: not this one, this is for invalid `$1`

**Pull request merged**